### PR TITLE
Add and update CRUD methods and tests

### PR DESCRIPTION
**Pull request by CCharlieLi**
@makara please have a check. I've noticed something weird, some methods of DataAccessObject in loopback-datasource-juggler do not match CRUD methods in CouchBase SDK, and this [doc](https://github.com/strongloop/loopback-datasource-juggler/blob/master/docs/datasource-connector.md) seems out of date as parameters of some CRUD methods have changed.

### CODE REVIEW

**Comment by makara** [comment]
You don't need to send me a new pull request; the previous one will be updated each time you make a commit.

**Comment by CCharlieLi** [comment]
Yes, just noticed that...

File: lib/couchbase3.js [hunk 1]
```diff
@@ -140,3 +145,231 @@ Couchbase.prototype.disconnect = function(callback) {
...
+    callback = noop;
+  }
+
+  // Handle with promise.
+  var promise;
+  var resolved = function(res) {
+    return callback(null, res);
+  };
+
+  // The cached promise.
```
Comment (makara) [inline lib/couchbase3.js]: I mean you could use `this.connect()` to get the cached promise and replace this entire block

File: lib/couchbase3.js [hunk 2]
```diff
@@ -140,3 +145,231 @@ Couchbase.prototype.disconnect = function(callback) {
...
+  if (promise == null) {
+    debug('No connections. Try to connect...');
+    promise = self.connect(callback);
+  }
+
+  //bucket id
+  var id = self.getIdValue(model, data);
+
+  promise = promise.then(function(bucket) {
+    Promise.promisifyAll(bucket);
```
Comment (makara) [inline lib/couchbase3.js]: Interesting. I doubt we want to use this but could you:

- Do this in `connect()`.
- Do a benchmark (on your local) and tell me which is faster, the callbacks or the "promisified" functions.

File: lib/couchbase3.js [hunk 3]
```diff
@@ -140,3 +145,231 @@ Couchbase.prototype.disconnect = function(callback) {
...
+    return bucket.insertAsync(id,data).then(function(result) {
```
Comment (makara) [inline lib/couchbase3.js]: Don't `then()` here; just return the promise that `insertAsync()` returns.

File: lib/couchbase3.js [hunk 4]
```diff
@@ -140,3 +145,231 @@ Couchbase.prototype.disconnect = function(callback) {
...
+      debug('Create successfully.');
+      return true;
+    });
+  }).catch(TypeError, function(e) {
```
Comment (makara) [inline lib/couchbase3.js]: Don't catch here; user will need to handle it themselves.

File: lib/couchbase3.js [hunk 5]
```diff
@@ -140,3 +145,231 @@ Couchbase.prototype.disconnect = function(callback) {
...
+    debug('Create failed:',e);
+  }).catch(function(e) {
```
Comment (makara) [inline lib/couchbase3.js]: Same, don't catch.

**Comment by makara** [comment]
> some methods of DataAccessObject in loopback-datasource-juggler do not match CRUD methods in CouchBase SDK

We are implementing things for `loopback-datasource-juggler` so we need to match what it expects.

> and this doc seems out of date as parameters of some CRUD methods have changed

Good catch. Please follow the code not the doc then. And if you got time you could send pull requests to fix the doc.

File: lib/couchbase3.js (hunk 2 above)
Comment (CCharlieLi) [inline lib/couchbase3.js]: Hi @makara ,
Thanks for your suggestions! I've tried both "callback" and "promisify", the cost of time for CRUD methods were almost the same, around 20ms.

**Comment by makara** [comment]
Hey @CCharlieLi 

- Are all the CRUD methods done?
- Can we have more tests, especially, more error tests?

**Comment by CCharlieLi** [comment]
Hi @makara ,
- Yes, I've finished CRUD methods that you can create, find, update and destroy document with document id.
- I just added three error tests to try to "Insert a existent document id", "find a non-existent document id" and "destroy a non-existent document id".
Thanks for your reminding.

**Comment by CCharlieLi** [comment]
- Add `Student` test data model: `{No: {type: String, id: true}, name: String, age: Number, emails:[String]}`
- Update CRUD tests: create more instances, test more queries, updates, destroy at the same time.

**Comment by makara** [comment]
Thanks. Merging.

**Pull request merged**